lib/executors.go

```diff
@@ -0,0 +1,341 @@
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package lib
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/loadimpact/k6/stats"
+	"github.com/loadimpact/k6/ui/pb"
+	"github.com/sirupsen/logrus"
+	null "gopkg.in/guregu/null.v3"
+)
+
+//TODO: remove globals and use some type of explicit dependency injection?
+//nolint:gochecknoglobals
+var (
+	executorConfigTypesMutex   sync.RWMutex
+	executorConfigConstructors = make(map[string]ExecutorConfigConstructor)
+)
+
+// ExecutionStep is used by different executors to specify the planned number
+// of VUs they will need at a particular time. The times are relative to their
+// StartTime, i.e. they don't take into account the specific starting time of
+// the executor, as that will be considered by the external execution executor
+// separately.
+//
+// A slice [{t1, v1}, {t2, v2}, {t3, v3}, ..., {tn, vn}] of execution steps
+// means that a executor will need 0 VUs until t1, it will need v1 number of
+// VUs from time t1 until t2, need v2 number of VUs from time t2 to t3, and so
+// on. t1 is usually 0, tn is usually the same as GetMaxDuration() and vn is
+// usually 0.
+//
+// Keep in mind that t(i) may be exactly equal to t(i+i), when there's an abrupt
+// transition in the number of VUs required by a executor. For example, the
+// variable-looping-vus executor may have 0-duration stages, or it may scale up
+// VUs in its last stage right until the end. These
+//
+// []ExecutionStep is also used by the ExecutorConfigMap, to represent the
+// amount of needed VUs among all executors, during the whole execution of a
+// test script. In that context, each executor's StartTime is accounted for and
+// included in the offsets.
+type ExecutionStep struct {
+	TimeOffset      time.Duration
+	PlannedVUs      uint64
+	MaxUnplannedVUs uint64
+}
+
+//TODO: make []ExecutionStep or []ExecutorConfig their own type?
+
+// ExecutorConfig is an interface that should be implemented by all executor config types
+type ExecutorConfig interface {
+	Validate() []error
+
+	GetName() string
+	GetType() string
+	GetStartTime() time.Duration
+	GetGracefulStop() time.Duration
+
+	// This is used to validate whether a particular script can run in the cloud
+	// or, in the future, in the native k6 distributed execution. Currently only
+	// the externally-controlled executor should return false.
+	IsDistributable() bool
+
+	GetEnv() map[string]string
+	GetExec() null.String //TODO: use interface{} so plain http requests can be specified?
```

Comment:
Minor: it's not immediately clear what `GetExec()` returns, especially if we switch to returning `interface{}`. Initially I thought it's short for "executor".

Consider renaming this to something more descriptive like `GetExecFunctionName` or `GetFunctionName`.